### DISCUSSION TITLE
Some added clarity on VBA...references and calling code across projects in tB

### DISCUSSION THREAD

**Discussion by mburns08109** [twinbasic/twinbasic#847] [General]
I pulled the following list from the references in a VBA project in an excel workbook I'm doing some work on.
I thought it might shed a little more light on the https://github.com/WaynePhillipsEA/twinbasic/issues/781 and https://github.com/WaynePhillipsEA/twinbasic/issues/818 issue discussions.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/MXB0310/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/MXB0310/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72" lang=EN-US style='tab-interval:.5in;
word-wrap:break-word'>

Name | Built-In | Type | IsBroken | guid | Fullpath | Description
-- | -- | -- | -- | -- | -- | --
VBA | True | 0 | False | {000204EF-0000-0000-C000-000000000046} | C:\Program Files\Common Files\Microsoft   Shared\VBA\VBA7.1\VBE7.DLL | Visual Basic For Applications
Excel | True | 0 | False | {00020813-0000-0000-C000-000000000046} | C:\Program Files\Microsoft   Office\root\Office16\EXCEL.EXE | Microsoft Excel 16.0 Object   Library
stdole | False | 0 | False | {00020430-0000-0000-C000-000000000046} | C:\Windows\System32\stdole2.tlb | OLE Automation
Office | False | 0 | False | {2DF8D04C-5BFA-101B-BDE5-00AA0044DE52} | C:\Program Files\Common Files\Microsoft   Shared\OFFICE16\MSO.DLL | Microsoft Office 16.0 Object Library
VBIDE | False | 0 | False | {0002E157-0000-0000-C000-000000000046} | C:\Program Files (x86)\Common   Files\Microsoft Shared\VBA\VBA6\VBE6EXT.OLB | Microsoft Visual Basic for Applications   Extensibility 5.3
Scripting | False | 0 | False | {420B2830-E718-11CF-893D-00A0C9054228} | C:\Windows\System32\scrrun.dll | Microsoft Scripting Runtime
PA3AddIn | False | 1 | False |   | C:\Users\MyUserID\AppData\Roaming\Microsoft\AddIns\PlnAsst3_Add-In.xlam |  
PersonalWB | False | 1 | False |   | C:\Users\MyUserID\AppData\Roaming\Microsoft\Excel\XLSTART\PERSONAL.XLSB |  

</body>

</html>

Dang that probably won't look so good on this page.
I'm just highlighting the differences between the Built-In references, "regular" COM/ActiveX references, and "Other workbook/other project" references in the VBE environment.

So, references may or may not have GUIDs. References may or may not have descriptions. FullPath is always present. a Shorter NAME is always present.

So, in terms of the ability to call code from AddIns and the like, there may be limits to the available information that tB may need to work with/work around.

Also note: the above references list (taken from a new blank, unsaved workbook file/project) can differ markedly from the references found in the personal.xlsb and the VBA Addin Workbook file/project - and they are NOT INHERITED from the referenced project/workbooks in the VBE (or when compiled) - so the ADODB reference in the Personal.xlsb's vbProject is not accessible.visible directly from the calling testxlsm's code modules and functions, but the called personalwb.module1.foo() can use the ADODB library without issue if it is locally referenced.

**Comment by mburns08109** [comment]
Another point to add - the last 2 references aren't strictly necessary in the VBE.
Code can be called like:

PersonalWB.module1.foo()  

...even without the added reference to the project.
from either within code, or from the VBA Immediate window.

...you all probably know that already, but if the workbook and VBA code modules that are open and "current"/active is "C;\temp\Test.xlsm"s modlule1.bas it puts a better picture on the execution context idea. because the code in PersonalWb.module1.foo() that references the ActiveWorkbook object will be referencing "C:\Temp\test.xlsm" as opposed to [PersonalWb] (and so on with activesheet, etc.).

I do this with some frequency at present and need to be able to rely on it.

**Reply by bclothier** [reply]
What you describe is basically defined by the VBA host. This is covered in [section 2.7.3](https://docs.microsoft.com/en-us/openspecs/microsoft_general_purpose_programming_languages/ms-vbal/a8700b6e-505b-4c7d-bd6a-1decbdf0ccfe). Also see [section 4.1](https://docs.microsoft.com/en-us/openspecs/microsoft_general_purpose_programming_languages/ms-vbal/4cd406c7-1ade-4522-8d7b-933183059ac7). Because it's implementation-defined, we cannot guarantee that the code that works in one host will still work in a non-VBA host or in a different VBA host unless we accurately simulate the same implementation as the original VBA host. Even among Microsoft Office hosts, they don't behave the same way or can behave in contradictory way. See: different implementations/behaviors of `Application.Run` in each VBA host. 

Cross-project references, however, is something that that is defined the same way for all VBA code and is covered in [section 5.6.10](https://docs.microsoft.com/en-us/openspecs/microsoft_general_purpose_programming_languages/ms-vbal/e3af3398-f090-40db-ade6-de3a93589c76). This, however, does not address any external identifiers which is provided by the host (c.f. [section 2.7.2](https://docs.microsoft.com/en-us/openspecs/microsoft_general_purpose_programming_languages/ms-vbal/8317bd66-b447-45c0-b16a-f5d3d0416d99) and [section 3.3.5.3](https://docs.microsoft.com/en-us/openspecs/microsoft_general_purpose_programming_languages/ms-vbal/74e5bf34-2a30-420c-9d3b-9341fa4c4c09))

**Comment by mburns08109** [comment]
@bclothier 

hah!
Thanks, Ben!
That [section 5.6.10](https://docs.microsoft.com/en-us/openspecs/microsoft_general_purpose_programming_languages/ms-vbal/e3af3398-f090-40db-ade6-de3a93589c76). link provides the answer to prove what I'd intuited all along for years now.
> § If the selected tier contains matches from multiple referenced projects, the matches from the project that has the highest reference precedence are retained and all others are discarded. < 

THIS is the root cause for the VBA errors when intrinsic functions like Left$() blow up on you suddenly when a broken reference is found in the chain of references.

Can Wayne make tB's resolver a fair bit smarter than VBAs in that regard?
